### ISSUE TITLE
[DDP + Dynamo] Tracing DDP AllReduce (Compiled DDP)

### ISSUE THREAD

**Issue by fegin** [pytorch/pytorch#109774]
### 🚀 The feature, motivation and pitch

**Background**
DistributedDataParallel (DDP) uses `Reducer` to bucket and issue  `allreduce` calls. The main entry point of `Reducer` is through the gradient hook of `AccumuldateGrad`. However, both `Reducer` and the gradient hook are not not traceable by Dynamo.  As a result, when using `torch.compile` to trace a DDP module, all the `allreduce` buckets will only be called at the end of the backend graph, resulting in zero overlapping between communication and backward computation. The current solution is to use `DDPOptimizer` to analyze the bucketing process of `Reducer` and breaks the forward graph into multiple ones so that each corresponding backward graph issues one `allreduce`.

With the recent development of CompiledAutograd, we would like to revisit tracing the whole DDP into a graph and capture all the `allreduce` in the backward graph.

**Approach**
The initial concept involves utilizing CompiledAutograd to trace the gradient hook registered by `Reducer` and properly record the `allreduce` calls. We don't need to trace the entire `Reducer` but only the bucketing process and `allreduce` calls. While CompiledAutograd could trace the gradient hook as a node, the gradient hook function must be traceable by Dynamo to be inlined in the graph challenge arises because the gradient hook o`Reducer` is written in C++ and not a torch op, Dynamo could not trace it. While it is possible to address this issue, the solutions can be difficult to maintain and handling rebucketing and graph breaks remains uncertain.

As an alternative approach, we propose the complete deactivation of `Reducer` and the registration of gradient hooks at the Python layer during DDP compilation. Unlike the gradient hook within `Reducer`, which has to manages bucketing, the Python gradient hook solely invokes the functional `allreduce` for the relevant gradient.  The Python gradient hook is simple to maintain. This tracing process will result in very bad communication performance but the `allreduce` calls are completely captured, allowing the subsequent optimization. We will implement the `allreduce` bucketing in Inductor.

**Pros/Cons**
The advantages of the proposed solution inherently support graph breaks. Since the graph does not contain any `allreduce` buckets, any graph break won't disturb the allreduce process, and the bucketing function in Inductor can continue functioning smoothly. Additionally, rebucketing is seamlessly managed by Dynamo/CompiledAutograd. When there are changes to the used parameters, Dynamo and CompiledAutograd will automatically recompile the model and trigger the rebucketing process.

A potential concern with the proposed solution lies in the compilation speed, particularly it can be slow for large models. The first iteration has to invoke unbucketing `allreduce`, which can be quite slow.

**Milestones**
1. Trace the gradient hook registered in the Python layer with `CompiledAutograd`
2. Trace DDP communication hook (the basic ones, like bf16_compress).
3. Disable `Reducer` when compiling.
4. Implement bucketing in Inductor.

We have validated milestones 1 and 2 and are working on milestone 3 and 4. After all the 4 milestones are done. We will have a MVP of fully traceable DDP.

cc., @wconstab, @jansel, @yf225, @voznesenskym, @ezyang, @animesh

### Alternatives

The proposed approach may also work with https://github.com/pytorch/pytorch/pull/109537/. We will verify with the PR after it is landed.

cc @H-Huang @awgu @wanchaol @fduwjj @wz337 @wconstab @d4l3k @pragupta @msaroufim @dcci @chauhang @penguinwu @mrshenli @pritamdamania87 @zhaojuanmao @satgera @rohan-varma @gqchen @aazzolini @osalpekar @jiayisuse @kwen2501 @XilunWu @ezyang @bdhirsh @anijain2305 @zou3519 @voznesenskym @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @chenyang78 @aakhundov @kadeng @kiukchung @lucasllc @tianyu-l @Xia-Weiwen

**Comment by pritamdamania87** [comment]
If we are implementing bucketing in inductor would it be possible for users to write passes where we can override the bucketing behavior and use our custom bucketing algorithm?

**Comment by fegin** [comment]
This is possible, though we will need to discuss what is the best UX that can avoid users from disrupting other optimizations that are not related to distributed in Inductor. cc., @yf225

**Comment by tbohutyn** [comment]
@fegin are the WIP changes available somewhere?

**Comment by IaroslavElistratov** [comment]
Is the plan for the `DDPOptimizer` to be replaced by this new approach or are these two intended to coexist together?

@fegin @wconstab

**Comment by wconstab** [comment]
Replace DDPOptimizer. Just have to get it done...

**Comment by zaptrem** [comment]
> Replace DDPOptimizer. Just have to get it done...

Why can't DDPOptimizer be used with SHARD_GRAD_OP/the Zero stage 1/2 modes of FSDP?

**Comment by wconstab** [comment]
> Why can't DDPOptimizer be used with SHARD_GRAD_OP/the Zero stage 1/2 modes of FSDP?

DDPOptimizer is designed for the case when Dynamo could capture a whole graph for .forward of your model, and introducing new (artificial) graph breaks helps overlap backwards compute + allreduce.

With FSDP, we currently graph-break on the FSDP logic itself during forward, and run communication operations out side the graph (during forward), so adding new graph-breaks via DDP optimizer is not helpful.

**Comment by fegin** [comment]
Update the status:

We have landed the PR to enable the Python level reducer to allow CompiledAutograd to trace through the allreduces. Another PR to do allreduce bucketing and reordering is also landed. We are verifying the performance and battle testing with some models. To try out the new CompiledDDP, one can use the following code:

```
# This has to be set before wrapping the model with DDP.
torch._dynamo.config.optimize_ddp = "python_reducer"

# compile the model as usual, here is just an example
model = torch.compile(DDP(model))

# Enable CompiledAutograd
with torch._dynamo.utils.maybe_enable_compiled_autograd(is_compiled):
    loss = model(input)
    # backward() call must be wrapped with the maybe_enable_compiled_autograd context manager.
    loss.backward()
```

cc., @lessw2020, @pritamdamania87

**Comment by YouJiacheng** [comment]
Hi! I encountered 2 problems:
1. It seems that we MUST "directly" compile the DDP wrapper, otherwise the gradients won't be synchronized.
I compiled the forward-backward function:
![image](https://github.com/user-attachments/assets/8419b628-69aa-4505-92ce-3de292a49069)
BUT the gradients were NOT synchronized if I didn't compile the DDP wrapper (i.e. without line 428).
![image](https://github.com/user-attachments/assets/92568b55-59d5-4b40-a11c-a13cb681f863)

2. If I compile the DDP wrapper, then I can't put the `weight` of my `CastedLinear` in `float32`. (`x` is in `bfloat16`)
![image](https://github.com/user-attachments/assets/b863db2d-0cd6-4f99-8814-5b4715ba55d3)
![image](https://github.com/user-attachments/assets/730a59bf-3809-47ed-9c59-33efa1351f2b)
